### PR TITLE
Implement Fold and Unfold

### PR DESCRIPTION
**Pull request by DuongQLee**
- Added [Fold](https://pytorch.org/docs/stable/generated/torch.nn.Fold.html) and [Unfold](https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html) op.
- Full benchmark result compared to ROCm <a href="https://docs.google.com/spreadsheets/d/1eeDMjad7lawFT0KGF-w1iALwCR14z-EYK5cjNy6m8e4/edit?usp=sharing">Here</a>
- Average performance:
<p>

| Op | Dtype | Direction | Time         |
|--------|-----------|-----------|--------------|
| Unfold | fp32      | fwd       | 16.43641979  |
|        | fp32      | bwd       | 1.458798342  |
|        | fp16      | fwd       | 15.83955361  |
|        | fp16      | bwd       | 1.459763543  |
|        | bfp16     | fwd       | 15.90593279  |
|        | bfp16     | bwd       | 1.455323877  |
| Fold   | fp32      | fwd       | 1.463731927  |
|        | fp32      | bwd       | 28.09828887  |
|        | fp16      | fwd       | 1.479315364  |
|        | fp16      | bwd       | 28.08557933  |
|        | bfp16     | fwd       | 1.47158127   |
|        | bfp16     | bwd       | 26.84660993  |
</p>

### CODE REVIEW

**Comment by iq136boy** [comment]
Erro logs:
[error_log.txt](https://github.com/user-attachments/files/16435015/error_log.txt)

**Comment by DuongQLee** [comment]
@iq136boy I can get the error log of cicd? Thank you

**Comment by DuongQLee** [comment]
@iq136boy Can I have the error log for Jenkins - Fp32 Hip All gfx90a? Thank you

**Comment by DuongQLee** [comment]
@iq136boy Can I have the error log for Jenkins - Fp32 Hip All gfx90a? Thank you.

File: src/kernels/MIOpenUnfold.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                                int LH,
+                                int LW,
+                                int kernel_size_h,
+                                int kernel_size_w,
+                                int stride_h,
+                                int stride_w,
+                                int padding_h,
+                                int padding_w,
+                                int dilation_h,
+                                int dilation_w,
```
Comment (CAHEK7) [inline src/kernels/MIOpenUnfold.cpp]: It limits the kernel applicability for tensors <2Gb

File: test/gtest/unfold.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    if(!MIOPEN_TEST_ALL ||
+       (env::enabled(MIOPEN_TEST_ALL) && env::value(MIOPEN_TEST_FLOAT_ARG) == "--float"))
+    {
+        RunTest();
+        Verify();
+    }
+    else
+    {
+        GTEST_SKIP();
+    }
```
Comment (CAHEK7) [inline test/gtest/unfold.cpp]: See https://github.com/ROCm/MIOpen/pull/3152#discussion_r1714048163

File: test/gtest/unfold.hpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+struct UnfoldTestCase
+{
+    size_t N;
+    size_t C;
+    size_t D;
+    size_t H;
+    size_t W;
+    std::vector<int32_t> kernelSize;
+    std::vector<int32_t> stride;
```
Comment (CAHEK7) [inline test/gtest/unfold.hpp]: While dimensions are `size_t`, strides are 32bit integers. Can the be consistent?

File: src/solver/fold/fold_forward.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+namespace solver {
+
+namespace fold {
+
+bool FoldFwd::IsApplicable(
+    [[maybe_unused]] const ExecutionContext& /*context*/,
+    [[maybe_unused]] const miopen::fold::FoldFwdProblemDescription& problem) const
+{
+    return true;
```
Comment (CAHEK7) [inline src/solver/fold/fold_forward.cpp:50]: Kernel does not support tensors >2Gb.

File: src/solver/fold/fold_forward.cpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+namespace miopen {
+
+namespace solver {
+
+namespace fold {
+
+bool FoldFwd::IsApplicable(
+    [[maybe_unused]] const ExecutionContext& /*context*/,
+    [[maybe_unused]] const miopen::fold::FoldFwdProblemDescription& problem) const
```
Comment (CAHEK7) [inline src/solver/fold/fold_forward.cpp]: Either
```C++
    const ExecutionContext& /*context*/,
```
or
```C++
    [[maybe_unused]] const ExecutionContext& context,
```

File: test/gtest/unfold.hpp [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+std::vector<UnfoldTestCase> UnfoldTestConfigs()
+{
+    // clang-format: off
+    return {
+        {2, 5, 0, 3, 4, {2, 3}, {1, 1}, {0, 0}, {1, 1}, true},
+        {1, 3, 0, 10, 12, {4, 5}, {1, 1}, {0, 0}, {1, 1}, true},
+        {11, 13, 0, 17, 19, {3, 3}, {3, 2}, {0, 0}, {1, 1}, true},
+        {11, 13, 0, 17, 19, {3, 3}, {1, 1}, {3, 2}, {1, 1}, true},
+        {11, 13, 0, 17, 19, {3, 3}, {1, 1}, {0, 0}, {3, 2}, true},
+        {11, 13, 0, 33, 37, {4, 3}, {2, 3}, {5, 2}, {3, 5}, true},
```
Comment (CAHEK7) [inline test/gtest/unfold.hpp:119]: It doesn't look like the algorithm rely on `isContiguous` and there is no `non-isContiguous` testcase to check it.

File: test/gtest/unfold.hpp [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        {
+            std::cout << "Error Input Tensor Lengths\n" << std::endl;
+            return std::vector<size_t>({0});
+        }
+    }
+
+    std::vector<size_t> ComputeStrides(std::vector<size_t> inputDim) const
+    {
+        if(!isContiguous)
+            std::swap(inputDim.front(), inputDim.back());
```
Comment (CAHEK7) [inline test/gtest/unfold.hpp:99]: Could you explain why do you swap input dimensions for `non-isContiguous` case?

File: test/gtest/unfold.hpp [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    void SetUp() override
+    {
+        auto&& handle = get_handle();
+        config        = GetParam();
+
+        std::vector<size_t> in_dims    = config.GetInput();
+        std::vector<size_t> in_strides = config.ComputeStrides(in_dims);
+
+        auto gen_value = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 100); };
+        [[maybe_unused]] auto gen_one = [&](auto...) { return 1; };
```
Comment (CAHEK7) [inline test/gtest/unfold.hpp]:
```suggestion
```

**Review by CAHEK7 — CHANGES_REQUESTED** [review]
Partial review.

File: test/gtest/unfold.hpp (hunk 6 above)
Comment (DuongQLee) [reply test/gtest/unfold.hpp:119]: I have added non contiguous test cases in commit https://github.com/ROCm/MIOpen/pull/3167/commits/85c1ee04b6737e6ac02251ae836b2a4c4f3cb101

File: test/gtest/unfold.hpp (hunk 7 above)
Comment (DuongQLee) [reply test/gtest/unfold.hpp:99]: For example, if we have tensor size {2, 4, 8, 16}, then contiguous strides is {512, 128, 16, 1}
If we swap the input strides and dim, then dim is {16, 4, 8, 2} and strides is {1, 16, 2, 64}
This will make the memory access non-contiguous but the size of the tensor still remains the same.

File: test/gtest/unfold.hpp (hunk 8 above)
Comment (DuongQLee) [reply test/gtest/unfold.hpp]: resolved in commit https://github.com/ROCm/MIOpen/pull/3167/commits/901d7b32448be557eeca4d3001ea648a71b477f0

File: src/solver/fold/fold_forward.cpp (hunk 5 above)
Comment (DuongQLee) [reply src/solver/fold/fold_forward.cpp]: resolve in commit https://github.com/ROCm/MIOpen/pull/3167/commits/918091d242ad9cad4eedd5696c425b6009d7b38a

File: test/gtest/unfold.hpp (hunk 7 above)
Comment (CAHEK7) [reply test/gtest/unfold.hpp:99]: Why can't it be just calculated in a reverse direction? Like {1, 2, 8, 64}. It is still the same total tensor size, but non-contiguous (i.e. transposed), but without any strange swapping.
Swapping looks suspicious.

**Comment by long10024070** [comment]
@iq136boy Would you send us the build log of this PR?

File: test/gtest/unfold.hpp (hunk 7 above)
Comment (DuongQLee) [reply test/gtest/unfold.hpp:99]: I have made a pytorch implementations of this method: https://colab.research.google.com/drive/1TEPXclDXDQ5cLHsoPpQmrBmVUvuPKDLe?usp=drive_link
As you can see, by swapping the front back dim (tensor dim is {16, 4, 8, 2}), the memory strides make the tensor incontiguous. 
image

**Comment by iq136boy** [comment]
Please check the error log:
[test_error_log.txt](https://github.com/user-attachments/files/16773897/test_error_log.txt)

**Review by iq136boy — COMMENTED** [review]
BLocked by following CI failure. Need to restart after the CI failure is solved.

[2024-09-23T13:54:42.218Z] Exception occurred: org.kohsuke.github.HttpException: {"message":"API rate limit exceeded for user ID 49319081. If you reach out to GitHub Support for help, please include the request ID

**Comment by hieule88** [comment]
I will handle this PR since @DuongQLee is now part of another team.

File: src/kernels/MIOpenUnfold.cpp (hunk 1 above)
Comment (hieule88) [reply src/kernels/MIOpenUnfold.cpp]: resolved

File: test/gtest/unfold.cpp (hunk 2 above)
Comment (hieule88) [reply test/gtest/unfold.cpp]: resolved

File: test/gtest/unfold.hpp (hunk 3 above)
Comment (hieule88) [reply test/gtest/unfold.hpp]: resolved, change them to uint64_t

File: src/solver/fold/fold_forward.cpp (hunk 4 above)
Comment (hieule88) [reply src/solver/fold/fold_forward.cpp:50]: I don't get this comment. Please give me more information

**Comment by hieule88** [comment]
@iq136boy Would you send us the build log of this PR?

**Review by iq136boy — DISMISSED** [review]
Please fix the following errors:
[error_log_3167.txt](https://github.com/user-attachments/files/17479436/error_log_3167.txt)

**Comment by hieule88** [comment]
> Please fix the following errors: [error_log_3167.txt](https://github.com/user-attachments/files/17479436/error_log_3167.txt)

Fixed it, @iq136boy Would you send us the latest build log of this PR?

**Review by iq136boy — COMMENTED** [review]
Error log:
[error_log (2).txt](https://github.com/user-attachments/files/17640577/error_log.2.txt)

**Review by iq136boy — COMMENTED** [review]
Error log:
[3167_log.txt](https://github.com/user-attachments/files/17820540/3167_log.txt)

**Comment by hieule88** [comment]
> Error log: [3167_log.txt](https://github.com/user-attachments/files/17820540/3167_log.txt)

I fixed this error. Please send me the newest error log of this PR.

**Review by tcgu-amd — APPROVED** [review]

**Comment by BradPepersAMD** [comment]
MIOpen is moving to the new monorepo setup and all older unmerged PR's are being closed.  Please re-open this as part of the new repo if these changes are still needed.
